yearn/utils.py

```diff
@@ -144,8 +143,9 @@ def __call__(self, *args, **kwargs):
 
 # cached Contract instance, saves about 20ms of init time
 _contract_lock = threading.Lock()
-_contract = lru_cache(maxsize=None)(Contract)
 
+# eth_retry must come second so lru_cache doesn't cache transient Exceptions.
+@lru_cache_with_exceptions(maxsize=None)
```

Comment:
TLDR; I think caching *all* exceptions regardless of the exception type or the reason of occurence is a bit dangerous and can lead to surprising results and longer debugging times. I don't recommend caching exceptions unless there's a verified performance issue when raising them in our usecase. There are some alternatives:

1. undo this change and don't cache exceptions at all, let the code fail for every block in the same way (assuming the error rate is low and performance is not impacted a lot by leaving as-is). This will always produce reproducable results with a minor performance overhead.
2. cache contract-specifc exceptions on the data layer (don't cache any other exceptions like system, network, io etc.) either by adapting the decorator or by creating an in-memory exclude dict and add the contract to that dict if an exception occured. Subsequent calls with another block would then first check if there's already a cached exception for a given contract and raise. This will only capture the first exception for the given code, shadowing any subsequent exceptions that might occur later until the process is restarted.
```python
a = '0xf00'
try:
  if a in failed_contracts:
    raise failed_contracts[a]
  c = Contract(a) 
except Exception as e:
  if str(e).endswith('has not been verified'): # fails because source is not verified
    failed_contracts[a] = e
  elif str(e)... # other conditions
    failed_contracts[a] = e
  else:
    raise e
```

Reply:
as discussed with Bob we exclude the caching of exception because it doesn't impact performance